collator/src/mempool/impls/std_impl.rs

```diff
@@ -334,3 +340,198 @@ impl MempoolAdapter for MempoolAdapterStdImpl {
+                    .boxed();
+
+            tokio::spawn(async move {
+                tokio::select! {
+                    () = &mut anchor_task => {}, // just poll
+                    // todo need a cancelation?
+                }
+            });
+
+            tracing::info!(target: tracing_targets::MEMPOOL_ADAPTER, "Mempool started");
+        }
+    }
+}
+
+impl MempoolAdapterSingleNodeImpl {
+    pub fn new(mempool_node_config: &MempoolNodeConfig, node_peer_info: &PeerInfo) -> Result<Self> {
+        let config_builder = MempoolConfigBuilder::new(mempool_node_config);
+        Ok(Self(Arc::new(Inner {
+            cache: Default::default(),
+            config: Mutex::new(ConfigAdapter {
+                builder: config_builder,
+                state_update_queue: Default::default(),
+                engine_session: None,
+            }),
+            node_peer_info: node_peer_info.clone(),
+            unprocessed_message_queue: Arc::new(Default::default()),
+            anchor_id: tokio::sync::Mutex::new(Cell::new(0)),
+            once_flag: std::sync::Once::new(),
+        })))
+    }
+}
+
+#[async_trait::async_trait]
+impl WrapperMempoolAdapter for MempoolAdapterSingleNodeImpl {
+    fn send_external(&self, message: Bytes) {
+        let mut queue = self.0.unprocessed_message_queue.lock().unwrap();
+        queue.push_back(message);
+    }
+    async fn update_config(
+        &self,
+        consensus_config: &Option<ConsensusConfig>,
+        genesis_info: &GenesisInfo,
+    ) -> Result<()> {
+        let mut config_guard = self.0.config.lock().await;
+        if let Some(consensus_config) = &consensus_config {
+            config_guard
+                .builder
+                .set_consensus_config(consensus_config)?;
+        } // else: will be set from mc state after sync
+
+        config_guard.builder.set_genesis(*genesis_info);
+        Ok::<_, anyhow::Error>(())
+    }
+}
+
+#[async_trait::async_trait]
+impl MempoolAdapter for MempoolAdapterSingleNodeImpl {
+    async fn handle_mc_state_update(
+        &self,
+        new_cx: crate::mempool::StateUpdateContext,
+    ) -> anyhow::Result<()> {
+        tracing::info!("call handle_mc_state_update");
+
+        tracing::debug!(
+            target: tracing_targets::MEMPOOL_ADAPTER,
+            full_id = %new_cx.mc_block_id,
+            "Received state update from mc block",
+        );
+
+        {
+            let cfg = &new_cx.consensus_config;
+            tracing::info!("handle_mc_state_update: consensus config={:?}", cfg);
+
+            let mut config_guard = self.0.config.lock().await;
+            config_guard
+                .builder
+                .set_consensus_config(&new_cx.consensus_config)?;
+        }
+
+        self.0.once_flag.call_once(|| {
+            let inner = self.0.clone();
+            tokio::spawn(async move {
+                let proccess_task = inner.start_process().boxed();
+                tokio::select! {
```

Comment:
Cancellation can be done here either by storing `AbortHandle` of this spawn somewhere, or by sharing `CancellationToken` from `tokio-util`

Reply:
`AbortHandle` was added.